Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp

```diff
@@ -0,0 +1,370 @@
+        // Look for hash prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Port:
+        // Look for pathname start
+        if (isNonSpecialPatternChar(m_tokenIndex, '/'))
+            changeState(URLPatternConstructorStringParserState::Pathname, 0);
+        else if (isSearchPrefix())
+            changeState(URLPatternConstructorStringParserState::Search, 1);
+        // Look for hash prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Pathname:
+        if (isSearchPrefix())
+            changeState(URLPatternConstructorStringParserState::Search, 1);
+        // Look for hash prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Search:
+        // Look for hash prefix
+        if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Hash:
+        // Do nothing
+        break;
+    case URLPatternConstructorStringParserState::Done:
+        // This step is never reached
+        ASSERT_NOT_REACHED();
+        break;
+    default:
+        break;
+    }
+}
+
+void URLPatternConstructorStringParser::performParse(ScriptExecutionContext& context)
+{
+    while (m_tokenIndex < m_tokenList.size()) {
+        m_tokenIncrement = 1;
+
+        if (m_tokenList[m_tokenIndex].type == URLPatternUtilities::TokenType::End) {
+            if (m_state == URLPatternConstructorStringParserState::Init) {
+                rewind();
+                if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+                    changeState(URLPatternConstructorStringParserState::Hash, 1);
+                else if (isSearchPrefix())
+                    changeState(URLPatternConstructorStringParserState::Search, 1);
+                else
+                    changeState(URLPatternConstructorStringParserState::Pathname, 0);
+
+                m_tokenIndex += m_tokenIncrement;
+                continue;
+            }
+            if (m_state == URLPatternConstructorStringParserState::Authority) {
+                rewind();
+                m_state = URLPatternConstructorStringParserState::Hostname;
+                m_tokenIndex += m_tokenIncrement;
+                continue;
+            }
+
+            changeState(URLPatternConstructorStringParserState::Done, 0);
+            break;
+        }
+
+        if (m_tokenList[m_tokenIndex].type == URLPatternUtilities::TokenType::Open) {
+            m_groupDepth += 1;
```

Comment:
@cdumez 

Should I change this to `++ m_groupDepth` instead? I recall that you had a preference for pre-increment but I don't remember why. 

In this situation, I actually think it's prettier to do `+= 1` but I don't have a strong preference.

Reply:
Okay I'll swap out all the `+=` for pre increment `++`